Initial prototype

Laukkala opened the pull request:
Prototype version
Main class serves as a demo for the prototype.

Features added:
Can create Process objects based on process IDs
Can check if Process is alive or not
Can get any file within the Process' /proc directory and return its contents in a List or a Map format Special object for a Process' stat-files which includes named keys for each value contained in the file. Special object for an OS's stat-file which includes named keys for each value contained in the file. Processes can list their Threads, which are also capable of providing information from their /proc files.

Laukkala committed:
Initial commit

Prototype version
Main class serves as a demo for the prototype.

Features added:
Can create Process objects based on process IDs
Can check if Process is alive or not
Can get any file within the Process' /proc directory and return its contents in a List or a Map format
Special object for a Process' stat-files which includes named keys for each value contained in the file.
Special object for an OS's stat-file which includes named keys for each value contained in the file.
Processes can list their Threads, which are also capable of providing information from their /proc files.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/teragrep/jos_01?pullRequest=2) <br/>All committers have signed the CLA.

Laukkala committed:
Fixed package names to conform to Maven file structure

Laukkala committed:
Fixed package names to conform to Maven file structure.
Made some fields final

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Main.java:
```diff
@@ -0,0 +1,109 @@
+        System.out.println("\nOS Specific files can be accessed via the proc() method");
+        System.out.println(os.proc("cpuinfo").rows());
+
+        os.pageSize();
```
possibly left over call?

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/OS.java:
```diff
@@ -0,0 +1,85 @@
+    private final File procDirectory;
+
+    public OS() {
+        procDirectory = new File("/proc");
```
please use new calls in secondary constructor only, this will allow the path of proc to be configurable, procfs can be mounted anywhere, while the default is to /proc by the https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/OS.java:
```diff
@@ -0,0 +1,85 @@
+import com.teragrep.jos_01.procfs.status.OSStat;
+import com.teragrep.jos_01.procfs.status.GenericStatus;
+
+public class OS {
```
perhaps it would rather be Linux because other operating systems have different kind of statistics

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/OS.java:
```diff
@@ -0,0 +1,85 @@
+        return pageSize;
+    }
+
+    public GenericStatus proc(String procFileName) {
```
GenericStatus sounds like a Reader class with an obfuscated name, is it so?

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/ProcFile.java:
```diff
@@ -0,0 +1,78 @@
+import java.io.*;
+import java.util.ArrayList;
+
+public class ProcFile extends File {
```
Perhaps RowFile, because there is nothing proc specific here

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,142 @@
+
+    public Process(long processId) {
+        this.processId = processId;
+        this.procDirectory = new File("/proc", Long.toString(processId));
```
use new only in secondary constructor so path to procfs is configurable.

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Task.java:
```diff
@@ -0,0 +1,135 @@
+                }
+            }
+            catch (NullPointerException npe) {
+                System.out.println("I/O Exception while attempting to access children of " + file.getPath());
```
please use name sub-directorie

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Task.java:
```diff
@@ -0,0 +1,135 @@
+                    procFileNames(nameList, child);
+                }
+            }
+            catch (NullPointerException npe) {
```
catching NPE is a no-go, please check for nulls if they are returned by the calls.

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Task.java:
```diff
@@ -0,0 +1,135 @@
+        return nameList;
+    }
+
+    public Long tid() {
```
please return primitive types instead of boxed ones (long instead of Long)

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/fields/stat/OSStatFields.java:
```diff
@@ -0,0 +1,50 @@
+ * Teragrep, the applicable Commercial License may apply to this file if you as
+ * a licensee so wish it.
+ */
+package com.teragrep.jos_01.procfs.fields.stat;
```
fields should reside in the same package as the class using them, it might be feasible to include the enum into the object so the responsibility of the knowledge imbued does not spread

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/status/ProcessStat.java:
```diff
@@ -0,0 +1,94 @@
+        this.rows = rows;
+        statistics = new LinkedHashMap<String, String>();
+        for (String row : rows) {
+            String[] fields = row.split(" (?![^()]*\\))"); // Regex matching spacebars except for within parentheses in case "comm" field contains spacebars
```
please use exact matching with Named Capture Groups such as 
```
(?<pid>\d+) (?<comm>\(.+\)) (?<state>\w) (?<a>\d+) (?<b>\d+) (?<c>\d+) (?<d>\d+) (?<e>\d+) (?<f>\d+) (?<g>\d+) (?<h>\d+) (?<i>\d+) (?<j>\d+) (?<k>\d+) (?<l>\d+) (?<m>\d+) (?<n>\d+) (?<o>\d+) (?<p>\d+) (?<q>\d+) (?<r>\d+) (?<s>\d+) (?<t>\d+) (?<u>\d+) (?<v>\d+) (?<w>\d+) (?<x>\d+) (?<y>\d+) (?<z>\d+) (?<a1>\d+) (?<b1>\d+) (?<c1>\d+) (?<d1>\d+) (?<e1>\d+) (?<f1>\d+) (?<g1>\d+) (?<h1>\d+) (?<j1>\d+) (?<k1>\d+) (?<l1>\d+) (?<m1>\d+) (?<n1>\d+) (?<o1>\d+) (?<p1>\d+) (?<q1>\d+) (?<r1>\d+) (?<s1>\d+) (?<t1>\d+) (?<u1>\d+) (?<v1>\d+) (?<w1>\d+) (?<x1>\d+)
```
regex101.com is a good tool with these

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/status/GenericStatus.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.Map;
+
+// Handles any proc file without setting any special formatting
+public class GenericStatus implements Status {
```
this sounds like a generic Reader class, files are quite unique so Generic wont do?

Laukkala committed:
OS: Added new primary constructor to LinuxOS allowing for configuration of /proc directory
OS: Renamed OS to LinuxOS as we only support Linux for now
GenericStatus: Removed GenericStatus as its only function is to read rows.
ProcFile: Renamed ProcFile to RowFile to better represent what the class does
Task: Replaced NullPointerException try-catch with a null check
Process: Added primary constructor to Process to make proc directory path configurable
OSStat: Added enum for OS stat fields
OSStatFields: Moved enum to OsStat class
ProcessStatFields: Moved enum to ProcessStat class
ProcessStat:Added enum for process stat fields
ProcessStat:Replaced regex pattern to use Named capture groups

Laukkala committed:
Removed proc-file specific "StatField" Enum classes, keys moved to the class rerpesenting the proc file.

Laukkala committed:
Added Statm object to represent a /proc/pid/statm file.

Laukkala committed:
Made Process.residentSetSize() use newly added Statm object instead of parsing the data itslef

Laukkala committed:
Added Meminfo and Vmstat objects to represent those OS-specific proc files.

Edited pageSize() to use these new files when calculating page size

Laukkala committed:
Added cpuTime() and cpuUsage() methods to Process
Added totalRam() and cpuCount() methods to LinuxOS

kortemik commented:
see also https://android.googlesource.com/platform/frameworks/base/+/android-6.0.0_r1/core/java/com/android/internal/os/ProcessCpuTracker.java

Laukkala committed:
Sysconf: JNA (Java Native Access) implementation of a very simple C program that uses sysconf(SC_CLK_TCK) to return the clock tick rate of the system.
This information is used to turn cpu-related information (reported in ticks) into seconds.

Laukkala committed:
Code formatting changes only

Laukkala committed:
Added CPUInfo and methods to get number of CPUs, Physical cores and Threads.

Renamed "statFields" to just "fields" within OSStat, Vmstat ProcessStat and Statm

StrongestNumber9 commented on src/test/test.java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,139 @@
+            Assert.assertNotNull(entry.getKey());
+        }
+
+        Process kthreadd = new Process(2);
```
Seems like two independent tests in one test?

StrongestNumber9 commented on pom.xml:
```diff
@@ -0,0 +1,238 @@
+  <properties>
+    <changelist>-SNAPSHOT</changelist>
+    <java.version>1.8</java.version>
+    <junit.version>4.12</junit.version>
```
Is old (4.x) junit used on purpose? Should be 5.x otherwise

Laukkala replied:
Changed to junit 5.9.2

StrongestNumber9 commented on src/test/test.java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,139 @@
+    @Test
+    public void constructorTest() {
+        Process process = new Process(1);
+        Process processString = new Process("1");
```
This test doesn't have any asserts

StrongestNumber9 reviewed (commented):
Test related notes

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/ProcessStat.java:
```diff
@@ -0,0 +1,98 @@
+        this.rows = rows;
+        statistics = new LinkedHashMap<String, String>();
+        for (String row : rows) {
+            Pattern pattern = Pattern.compile("(?<pid>\\d+) (?<comm>\\(.+\\)) (?<state>\\w) (?<ppid>\\d+) (?<pgrp>\\d+) (?<session>\\d+) (?<ttynr>\\d+) (?<tpgid>-?\\d+) (?<flags>\\d+) (?<minflt>\\d+) (?<cminflt>\\d+) (?<majflt>\\d+) (?<cmajflt>\\d+) (?<utime>\\d+) (?<stime>\\d+) (?<cutime>\\d+) (?<cstime>\\d+) (?<priority>\\d+) (?<nice>\\d+) (?<numthreads>\\d+) (?<itrealvalue>\\d+) (?<starttime>\ ...
```
Should this be outside the loop? Otherwise creating it over and over again with same arguments

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Meminfo.java:
```diff
@@ -0,0 +1,102 @@
+        this.rows = rows;
+        statistics = new LinkedHashMap<String, String>();
+        for (String row : rows) {
+            Pattern pattern = Pattern.compile("(\\w*): *(\\d*)");
```
Should this be outside the loop?

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,175 @@
+
+    // Only the OS kernel can write or delete files from /proc, so if the process ID directory exists, the process is alive.
+    public boolean isAlive() {
+        if (procDirectory.exists()) {
```
Shouldn't this just be `return procDirectory.exists()`?

StrongestNumber9 reviewed (commented):
Potential performance related notes

Laukkala committed:
Added Uptime object to represent OS /proc/uptime file

Laukkala committed:
Removed generic proc() method from LinuxOS, Process and Task. Use file-specific methods to access these files instead.

Laukkala committed:
Added unit tests for OS status objects

Laukkala committed:
Added JNA implementation of _SC_CLK_TCK

Laukkala committed:
Whitespace and code formatting changes only

Laukkala committed:
Updated tests to use Junit5 instead of Junit4

Laukkala committed:
Added assertions to constructorTest to verify that objects have been initialized properly.

Laukkala committed:
Moved Pattern matching outside of a loop to improve performance.

Laukkala committed:
Fixed Meminfo not matching keys with parentheses properly

Laukkala committed:
Spotless formatting changes

StrongestNumber9 commented on README.adoc:
```diff
@@ -23,13 +28,21 @@ See the official documentation on https://docs.teragrep.com[docs.teragrep.com].
 
 == How to [compile/use/implement]
 
+Run "mvn clean install" successfully before use.
```
Mention java version explicitly if there are any limitations between java lts (8/11/17/21) versions

Laukkala replied:
Added java 1.8 to readme

StrongestNumber9 commented on pom.xml:
```diff
@@ -0,0 +1,357 @@
+  <properties>
+    <changelist>-SNAPSHOT</changelist>
+    <java.version>1.8</java.version>
+    <junit.version>4.12</junit.version>
```
Seems like junit version is explicitly set in the dependencies but this property still exists

Laukkala replied:
Removed property

StrongestNumber9 commented on pom.xml:
```diff
@@ -0,0 +1,357 @@
+      </plugin>
+      <plugin>
+        <groupId>org.codehaus.mojo</groupId>
+        <artifactId>exec-maven-plugin</artifactId>
```
Can you add explicit version number

StrongestNumber9 commented on pom.xml:
```diff
@@ -0,0 +1,357 @@
+      </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-jar-plugin</artifactId>
```
Can you add explicit vesrion number

StrongestNumber9 commented on pom.xml:
```diff
@@ -0,0 +1,357 @@
+            <manifest>
+              <addClasspath>true</addClasspath>
+              <!-- <classpathPrefix>lib</classpathPrefix> -->
+              <!-- <mainClass>test.org.Cliente</mainClass> -->
```
Remove commented out code

Laukkala replied:
Removed

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/LinuxOS.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    // Estimates page size in kB.
+    float pageSize() {
```
Is this intentionally without private/public flag?

Laukkala replied:
Added as public

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/LinuxOS.java:
```diff
@@ -0,0 +1,137 @@
+        Sysconf sysconf = new Sysconf();
+        long clkTck = sysconf.main();
+        if (clkTck == -1) {
+            throw new IOException("");
```
Add informative message

Laukkala replied:
Added message specifying what went wrong

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/Main.java:
```diff
@@ -0,0 +1,132 @@
+/*
```
This project reads like it is supposed to be a library but this file suggests to a standalone jar.

Laukkala replied:
Main class is more of a demo in this case, I suppose a library doesn't even need a main class, since this won't be used from the command line

kortemik replied:
please move to ReadmeTest.java under tests or so?

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,173 @@
+        else {
+            File[] subdirectories = file.listFiles();
+            if (subdirectories == null) {
+                System.out
```
Use proper logger unless there is a reason it can't be used

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,173 @@
+            nameList.add(file.getPath().replace(procDirectory.getPath(), ""));
+        }
+        else {
+            File[] subdirectories = file.listFiles();
```
Is this subdirectories used anywhere except for permission test?

Laukkala replied:
Realized I can use the same subdirectories list in the loop after permission check to remove an unnecessary file.listFiles() call

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,173 @@
+                return nameList;
+            }
+            for (File child : file.listFiles()) {
+                procFileNames(nameList, child);
```
Does the it modify the current list or is this a noop?

Laukkala replied:
This will modify the list, recursively adding filenames of all child files to the list, finally returning a list containing all filenames to which we have permission to access.

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/RowFile.java:
```diff
@@ -0,0 +1,78 @@
+        ArrayList<String> rows = new ArrayList<String>();
+        try {
+            BufferedReader br = new BufferedReader(new FileReader(this));
+            {
+                String line;
+                while ((line = br.readLine()) != null) {
+                    rows.add(line);
+                }
+            }
```
Unnecessary `{  }` block?

Laukkala replied:
Removed

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/RowFile.java:
```diff
@@ -0,0 +1,78 @@
+        }
+        catch (FileNotFoundException notFoundException) {
+            System.out.println("File \"" + this.getPath() + "\" not found!");
+            System.out.println("Reason: " + notFoundException.getMessage());
```
Use proper logger if possbile, also is returning an empty `ArrayList` expected when encountering issues?

Laukkala replied:
Changed this to re-throw a new exception which includes the additional information presented by the System.out.println, instead of printing the error details to console

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/RowFile.java:
```diff
@@ -0,0 +1,78 @@
+            System.out.println("Reason: " + notFoundException.getMessage());
+        }
+        catch (IOException ioException) {
+            System.out.println("Failed to read file \"" + this.getName() + "\" due to an I/O exception!");
```
Use proper logger if possbile, also is returning an empty `ArrayList` expected when encountering issues?

Laukkala replied:
Changed this to re-throw a new exception which includes the additional information presented by the System.out.println, instead of printing the error details to console

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/Sysconf.java:
```diff
@@ -0,0 +1,72 @@
+    public interface LibSysconf extends Library {
+
+        LibSysconf INSTANCE = (LibSysconf) Native
+                .load("/lib/opt/Fail-Safe/jos_01/lib/sysconf/sysconf.so", LibSysconf.class);
```
Is this file guaranteed to exist, be valid and readable?

Laukkala replied:
File is generated when installing the project with "mvn install", but it's possible that user deletes or can't access the file. Added exception handling to at least inform user of these issues

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/Sysconf.java:
```diff
@@ -0,0 +1,72 @@
+    public long main() throws IOException {
+        long clk_tck = LibSysconf.INSTANCE.main();
+        if (clk_tck == -1) {
+            throw new IOException("Could not get system clock tick rate! sysconf(_SC_CLK_TCK) encountered an error!");
```
Is IOException correct exception to throw from syscall?

Laukkala replied:
Maybe not, perhaps a RuntimeException instead?

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/Task.java:
```diff
@@ -0,0 +1,131 @@
+
+    // Only the OS kernel can write or delete files from /proc, so if the process ID directory exists, the process is alive.
+    public boolean isAlive() {
+        if (procDirectory.exists()) {
```
This could be just `return procDirectory.exists()`?

Laukkala replied:
Fixed

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Cpuinfo.java:
```diff
@@ -0,0 +1,381 @@
+    }
+
+    public int cpuPhysicalCoreCount() {
+        Map<Integer, Integer> coreCounts = new HashMap<>();
+        for (Processor processor : processors) {
+            coreCounts.put(processor.physical_id, processor.cpu_cores);
+        }
+        int coreCount = 0;
```
I think this function could be optimized down to
```
int coreCount = 0;
for(Processor processor : processors) {
    coreCount += proecssor.cpu_cores
}
return coreCount;
```

Laukkala replied:
This could lead to duplication, because cpuinfo's processors are logical processors, and each reports the number of physical cpu cores the associated physical processor has.

The first loop goes through every logical processor, and adds each physical processor's cpu_cores. Since the keys in Map have to be unique, adding a cpu_cores value to an already existing key just updates that key's value instead of adding them together, handling the duplication issue.

We could just read the first occurrence of cpu_cores and call it a day, but it would be incorrect if the system has multiple physical CPU's installed.

Perhaps Cpuinfo.Processor should be renamed to Cpuinfo.LogicalProcessor

StrongestNumber9 replied:
Sounds reasonable

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Cpuinfo.java:
```diff
@@ -0,0 +1,381 @@
+    }
+
+    public int cpuCount() {
+        ArrayList<Integer> physicalIds = new ArrayList();
+        for (Processor processor : processors) {
+            physicalIds.add(processor.physical_id);
+        }
+        Map<Integer, Integer> distinct = new HashMap<>();
+        for (int physicalId : physicalIds) {
+            distinct.put(physicalId, distinct.getOrDefault(physicalId, 0) + 1);
+        }
+        return distinct.size();
```
Same as above

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Cpuinfo.java:
```diff
@@ -0,0 +1,381 @@
+
+    public void printStatistics() {
+        for (Map.Entry<String, String> statistic : statistics.entrySet()) {
+            System.out.print(statistic.getKey() + ": ");
```
Use logger instead of stdout?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Cpuinfo.java:
```diff
@@ -0,0 +1,381 @@
+        return processors.size();
+    }
+
+    public Map<String, String> statistics() {
```
What happens if I run this function in a loop? Seems like the `statistics` will bloat until memory runs out?

Laukkala replied:
The size of statistics shouldn't increase with subsequent calls, since the keys of the Map don't change in \proc\cpuinfo. Subsequent calls would just replace the values for each pre-existing key.

Regardless, added a call to statistics.clear(), in case it's possible that the number of logical cpus changes in \proc\cpuinfo. This could lead to stale data being stuck in the Map.

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Cpuinfo.java:
```diff
@@ -0,0 +1,381 @@
+    }
+
+    public void printTimestamp() {
+        System.out.println(timestamp);
```
Use logger instead of stdout?

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Meminfo.java:
```diff
@@ -0,0 +1,160 @@
+            Matcher matcher = pattern.matcher(row);
+            if (matcher.find()) {
+                for (int i = 0; i < matcher.groupCount(); i++) {
+                    statistics.put(matcher.group(1), matcher.group(2));
```
This sounds like it is adding duplicates? Unless something like this was the intention?
```
while(matcher.find()) {
    statistics.put(matcher.group(1), matcher.group(2));
}
```

Laukkala replied:
Removed unnecessary for-loop

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Meminfo.java:
```diff
@@ -0,0 +1,160 @@
+
+    public void printStatistics() {
+        for (Map.Entry<String, String> statistic : statistics.entrySet()) {
+            System.out.print(statistic.getKey() + ": ");
```
Should logger be used instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Meminfo.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    public void printTimestamp() {
+        System.out.println(timestamp);
```
Should logger be used instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/OSStat.java:
```diff
@@ -0,0 +1,97 @@
+        this.rows = rows;
+        statistics = new LinkedHashMap<String, String>();
+        for (int i = 0; i < rows.size(); i++) {
+            {
+                String[] fieldArray = rows.get(i).split(fields.values()[i].name());
+                statistics.put(fields.values()[i].toString(), fieldArray[1]);
+            }
```
Extra `{ }` block?

Laukkala replied:
Removed

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/OSStat.java:
```diff
@@ -0,0 +1,97 @@
+
+    public void printStatistics() {
+        for (Map.Entry<String, String> statistic : statistics.entrySet()) {
+            System.out.print(statistic.getKey() + ": ");
```
Use logger instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/OSStat.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    public void printTimestamp() {
+        System.out.println(timestamp);
```
Use logger instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/ProcessStat.java:
```diff
@@ -0,0 +1,157 @@
+
+    public void printStatistics() {
+        for (Map.Entry<String, String> statistic : statistics.entrySet()) {
+            System.out.print(statistic.getKey() + ": ");
```
Use logger instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/ProcessStat.java:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    public void printTimestamp() {
+        System.out.println(timestamp);
```
Use logger instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/ProcessStat.java:
```diff
@@ -0,0 +1,157 @@
+                );
+        for (String row : rows) {
+            Matcher matcher = pattern.matcher(row);
+            if (matcher.find()) {
+                for (int i = 0; i < matcher.groupCount(); i++) {
+                    statistics
+                            .put(fields.values()[i].name(), matcher.group(fields.values()[i].name().replace("_", "")));
+                }
+            }
```
I *THINK* this might be same as the other `while(matcher.find())` comment but I am not exactly sure.

Laukkala replied:
This file has different formatting where a single row can contain multiple values that need parsing so the inner loop is needed here

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Statm.java:
```diff
@@ -0,0 +1,109 @@
+                );
+        for (String row : rows) {
+            Matcher matcher = pattern.matcher(row);
+            if (matcher.find()) {
+                for (int i = 0; i < matcher.groupCount(); i++) {
+                    statistics
+                            .put(fields.values()[i].name(), matcher.group(fields.values()[i].name().replace("_", "")));
+                }
+            }
```
This *MIGHT* be the same as the other `while(matcher.find())` comment but I am not sure.

Laukkala replied:
This one has the same formatting with multiple values in a single row

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Statm.java:
```diff
@@ -0,0 +1,109 @@
+
+    public void printStatistics() {
+        for (Map.Entry<String, String> statistic : statistics.entrySet()) {
+            System.out.print(statistic.getKey() + ": ");
```
Use logger instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Statm.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    public void printTimestamp() {
+        System.out.println(timestamp);
```
Use logger instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Uptime.java:
```diff
@@ -0,0 +1,102 @@
+        Pattern pattern = Pattern.compile("(?<uptimeSeconds>\\d+.\\d+) (?<combinedCpuCoreIdleTimeSeconds>\\d+.\\d+)");
+        for (String row : rows) {
+            Matcher matcher = pattern.matcher(row);
+            if (matcher.find()) {
```
This *MIGHT* be the same as the other `while(matcher.find())` comment but I am not sure.

Laukkala replied:
This one has the same formatting with multiple values in a single row

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Uptime.java:
```diff
@@ -0,0 +1,102 @@
+
+    public void printStatistics() {
+        for (Map.Entry<String, String> statistic : statistics.entrySet()) {
+            System.out.print(statistic.getKey() + ": ");
```
Use logger instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Uptime.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    public void printTimestamp() {
+        System.out.println(timestamp);
```
Use logger instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Vmstat.java:
```diff
@@ -0,0 +1,234 @@
+        Pattern pattern = Pattern.compile("(.*) (.*)");
+        for (String row : rows) {
+            Matcher matcher = pattern.matcher(row);
+            if (matcher.find()) {
```
See the first `matcher.find()` comment

Laukkala replied:
This one is a newline-separated file and has a redundant for-loop! Removed.

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Vmstat.java:
```diff
@@ -0,0 +1,234 @@
+
+    public void printStatistics() {
+        for (Map.Entry<String, String> statistic : statistics.entrySet()) {
+            System.out.print(statistic.getKey() + ": ");
```
Use logger instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/main/java/com/teragrep/jos_01/procfs/status/Vmstat.java:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    public void printTimestamp() {
+        System.out.println(timestamp);
```
Use logger instead?

Laukkala replied:
Replaced stdout with a logger

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,212 @@
+
+        LinuxOS os = new LinuxOS();
+        OSStat stat = os.stat();
+        Assertions.assertEquals(12, stat.statistics().size());
```
Does this depend on your processor having only 4 cores?

Laukkala replied:
Yes, added a counter to test to handle different number of cpus

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,212 @@
+    public void cpuInfoTest() {
+        LinuxOS os = new LinuxOS();
+        Cpuinfo cpuinfo = os.cpuinfo();
+        Assertions.assertEquals(os.cpuThreadCount() * 27, cpuinfo.statistics().size());
```
Where does this *27 come from?

Laukkala replied:
Each logical processor (or a cpu thread) should give 27 different pieces of information in \proc\cpuinfo, which should all be listed in cpuinfo.statistics. So there should be exactly 27*cpuThreadCount unique fields in cpuinfo.statistics

StrongestNumber9 replied:
Add that as a comment for future devs

Laukkala replied:
Added comment to test

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,212 @@
+                "power management"
+        };
+
+        System.out.println(cpuinfo.statistics());
```
Use logger instead?

Laukkala replied:
Changed stdouts to use a Logger

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,212 @@
+                "DirectMap2M",
+                "DirectMap1G"
+        };
+        System.out.println(meminfo.statistics());
```
Use logger instead?

Laukkala replied:
Changed stdouts to use a Logger

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,212 @@
+    @Test
+    public void highLevelMethodsTest() {
+        LinuxOS os = new LinuxOS();
+        Assertions.assertEquals(1, os.cpuCount());
```
Will fail on non-single socket machines. This whole test will fail on other setups. Something like https://docs.oracle.com/javase%2F8%2Fdocs%2Fapi%2F%2F/java/lang/Runtime.html#availableProcessors-- could MAYBE be used as a workaround but I think it also contains thread count instead of logical core count.

Laukkala replied:
Used Runtime.availableProcessors() where applicable. Changed other tests to simply check that a positive integer gets returned, whatever the number of cpus the user's system ends up being

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,192 @@
+    public void constructorTest() {
+        Process process = new Process(1);
+        Process processString = new Process("1");
+        Assertions.assertNotNull(process.isAlive());
```
Use `Assertions.assertTrue()` as that is more correct to use, that method should never return null

Laukkala replied:
Fixed

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,192 @@
+    @Test
+    public void procFileNamesTest() {
+        Process process = new Process(1);
+        ArrayList<String> fileNames = process.availableProcFiles();
```
Is this machine specific value?

Laukkala replied:
Not sure if this value changes between machines, but it might. Will need to look for another way to test this

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,192 @@
+        Process process = new Process(1);
+        ProcessStat stat = process.stat();
+        for (String key : expectedKeys) {
+            System.out.println(key + " " + stat.statistics().get(key));
```
Use logger instead?

Laukkala replied:
Changed stdouts to use a Logger

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,192 @@
+        Process process = new Process(1);
+        Statm statm = process.statm();
+        for (String key : expectedKeys) {
+            System.out.println(key + " " + statm.statistics().get(key));
```
Use logger instead?

Laukkala replied:
Changed stdouts to use a Logger

StrongestNumber9 reviewed (changes requested):
Added bunch of notes. Add comments to conversations and resolve them if you fix any

Laukkala committed:
Updated Readme with Java version
Removed mentions of junit 4 from pom
Added error message to LinuxOS.cpuTicksPerSecond
Removed unnecessary file.fileList() call in Process
Removed unnecessary {} block from RowFile
RowFile now re-throws an exception instead of returning an empty list when a file read fails or a file is not found. Added Exception signatures to all related methods
Added exception to Sysconf in case sysconf.so file has not been generated and changed exception types to be more descriptive
Added a Statistics.clear() call to cpuInfo in case the number of logical cpus change
Removed unnecessary loop from MemInfo and VMstat
Removed unnecessary {} block from OSStat

Laukkala committed:
Replaced prints to stdout with a Logger

Laukkala committed:
Added test for process.isAlive()

Laukkala committed:
Added SysconfInterface which includes a Fake object for testing. SysconfInterface.Fake does not call the native C code, just returns a hardcoded "100". Should not be used in normal operation

Laukkala committed:
Made Process.cpuUsage support SysconfInterface.Fake()

Laukkala committed:
Added tests to ProcessTest, should cover all current methods.

Laukkala committed:
Added tests to LinuxOSTest, should cover all current methods. Edited tests to be system-independent

Laukkala committed:
Added tests to LinuxOSTest, should cover all current methods. Edited tests to be system-independent

Laukkala committed:
Fixed LinuxOSTest.statTest to be system independent

Laukkala committed:
Formatting changes

Laukkala committed:
Changed CPU usage tests to be more reliable

Laukkala committed:
Added version numbers to dependencies

kortemik commented on pom.xml:
```diff
@@ -0,0 +1,344 @@
+                <LDFLAGS>-pie -Wl,-z,relro -Wl,-z,now</LDFLAGS>
+                <LD_LIBRARY_PATH></LD_LIBRARY_PATH>
+                <LIBS>-ldl -lrt</LIBS>
+                <CFLAGS>-ggdb -fpie -fPIC -std=c99 -I/usr/include/sysconf -DHAVE_ATOMIC_BUILTINS64 -D_DEFAULT_SOURCE</CFLAGS>
```
-I/usr/include/sysconf is unnecessary because it does not exist as a directory?

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/LinuxOS.java:
```diff
@@ -0,0 +1,147 @@
+        return cpuTicksPerSecond(new Sysconf());
+    }
+
+    public long cpuTicksPerSecond(SysconfInterface sysconf) throws IOException {
```
why an argument is required? should this be a separate object or just instantiate new SysconfImpl() in a secondary constructor and call primary constructor of LinuxOS from that, as test cases may require a fake?

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,185 @@
+    private final Logger LOGGER = LoggerFactory.getLogger(Process.class);
+
+    public Process(String processId) {
+        this(Integer.parseInt(processId));
```
Long.parseLong because primary expects long

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,185 @@
+        return new Statm(rows);
+    }
+
+    public ArrayList<String> availableProcFiles() {
```
is there a use case for this method?

Laukkala replied:
Not anymore, since we are no longer using generic proc(fileName) calls, Removed these methods.

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,185 @@
+        else {
+            File[] subdirectories = file.listFiles();
+            if (subdirectories == null) {
+                LOGGER
```
throw IOException instead of logging

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,185 @@
+
+    public ArrayList<Task> tasks() {
+        ArrayList<Task> tasks = new ArrayList<Task>();
+        for (File taskDirectory : new File(procDirectory, "task").listFiles()) {
```
listFiles() may return null which must be handled properly: IOException when it's null.

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,185 @@
+    // Prints RSS in kB
+    public float residentSetSize() throws IOException {
+        Statm statm = statm();
+        String rssPages = statm.statistics().get("resident");
```
is getter necessary perhaps code should be refactored into an object and getter removed.

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    public float cpuUsage() throws IOException {
+        return cpuUsage(new Sysconf());
```
sysconf could be class field initialized in secondary constructor

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,185 @@
+        int cpuCount = os.cpuCount();
+        long cpuTicksPerSecond = os.cpuTicksPerSecond(sysconf);
+
+        float OSUpTime = Float.parseFloat(os.uptime().statistics().get("uptimeSeconds"));
```
x marks the spot

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/status/Statm.java:
```diff
@@ -0,0 +1,114 @@
+        timestamp = LocalDateTime.now();
+    }
+
+    public Map<String, String> statistics() {
```
usage of this map seems to be usually Map.get("metric name"), however as in the regex the metric names and their type is already known (enum-key, long value) the map type is not a proper one. Perhaps it would be wise to rework this into an object that has dedicated methods to access the appropriate items such as statm.size() statm.resident() and if ever there is need for map presentation it would be toMap()

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/status/Statm.java:
```diff
@@ -0,0 +1,114 @@
+
+    public Statm(ArrayList<String> rows) {
+        this.rows = rows;
+        statistics = new LinkedHashMap<String, String>();
```
new is only allowed in secondary constructor, main constructor must not have it. Perhaps the RowFile should be decorated with Statm instead of having rows as argument.

kortemik replied:
please consider this approach for other implementations as well

kortemik reviewed (changes requested):
partial review, however please check the points and apply to other places with similar issues. please review instructions from the EO vol 1 in case necessary

Laukkala committed:
Removed unnecessary argument from C compilation script

Laukkala committed:
Process.java: Secondary constructor of Process uses longs instead of ints

Laukkala committed:
Added check for nulls from File.listFiles() within tasks() and replaced use of Loggers with Exceptions

Laukkala committed:
Removed the required argument from cpuTime() and cpuUsage() by associating a LinuxOS object with each Process object.
LinuxOS is now associated with a Sysconf object, which can be a fake for tests.

Laukkala committed:
Removed the required argument from cpuTime() and cpuUsage() by associating a LinuxOS object with each Process object.
Process is now associated with a LinuxOS object, removing calls to new LinuxOS() from residentSetSize(), cpuTime() and cpuUsage()
LinuxOS is now associated with a Sysconf object, which can be a fake for tests.

Laukkala committed:
Removed readProcFile() methods from LinuxOS, Task and Process as they are redundant and edited Status object constructors to accept RowFiles instead, cutting out the unnecessary call to readProcFile()

Laukkala committed:
Removed new keywords from outside of secondary constructors in RowFile

Laukkala committed:
Refactored different status objects to contain specific methods for each piece of information available for them.

Created Composable Decorators for parsing RowFiles in different ways. You can delimit the text by character, filter out non-digits, trim, etc by using the relevant decorator.

Created constructors for Status objects so you can make them with composable decorators

Edited tests accordingly.

Laukkala committed:
Removed AvailableProcFiles() as there is no use case for them anymore.

Laukkala committed:
Spotless formatting changes

Laukkala committed:
Refactored "new" keyword into secondary consturctors in Status objects.

Laukkala committed:
Refactored "new" keyword into secondary consturctors in Status objects.

Laukkala committed:
Refactored all calls from main() class into tests and removed main class.

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,244 @@
+
+    // Uptime should have 2 fields, and the values should be positive floats.
+    @Test
+    public void upTimeTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,244 @@
+    // Stat should have 7 fields in addition to cpu lines, and all expected keys should be present.
+    // The number of cpu lines is one aggregate line + one line for each logical core. The number of logical cores should be the same as reported in /proc/cpuinfo
+    @Test
+    public void statTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,244 @@
+    // Check that vmStat instantiates every required keys
+    // vmstat might have varying number of keys
+    @Test
+    public void vmStatTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,244 @@
+
+    // CPUInfo should have 27 fields for every logical cpu (cpuThread) listed.
+    @Test
+    public void cpuInfoTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,244 @@
+
+    // MemInfo should have exactly 34 fields, and object should be initialized with non-negative integer values.
+    @Test
+    public void memInfoTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,244 @@
+    // CpuThreadCount should either match or be greater than the number of physical cores
+    // CpuThreadCount and number of physical cores should be greater or equal to the number of CPUs installed
+    @Test
+    public void highLevelMethodsTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,331 @@
+        }
+    }
+
+    public ArrayList<Integer> getPidByCommand(String command) throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,331 @@
+
+    // Kthreadd process (pid = 1) should have 52 fields in its Stat file, and it should have (systemd) as comm value and 1 as pid value in the file. None of the fields should be null
+    @Test
+    public void readSystemdFileTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,331 @@
+
+    // Kthreadd process (pid = 2) should have 52 fields in its Stat file, and it should have (kthreadd) as comm value and 2 as pid value in the file. None of the fields should be null
+    @Test
+    public void readKthreaddFileTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,331 @@
+
+    // After starting a sleeping process in a new thread, isAlive should return true. After killing the process, it should return false.
+    @Test
+    public void isAliveTest() throws IOException, NoSuchFieldException, IllegalAccessException, InterruptedException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,331 @@
+            kill.waitFor();
+        }
+        catch (InterruptedException inter) {
+            LOGGER.warn("Kill interrupted!");
```
This should not happen, assertion it instead

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,331 @@
+
+    // The number of tasks should be equal to the number of threads in /proc/pid/stat. Test for both systemd and the current JVM
+    @Test
+    public void correctNumberOfTasksTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,331 @@
+
+    // Stat status object should contain all of the listed fields.
+    @Test
+    public void statTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,331 @@
+
+    // Timestamps should be available and should be different if the proc method is called later
+    @Test
+    public void timestampTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,331 @@
+
+    // Statm status object should contain all of the listed fields.
+    @Test
+    public void statmTest() throws IOException {
```
Don't throw exceptions in tests

StrongestNumber9 reviewed (changes requested):
Theres quite a lot of `throws exception` used which is not good

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/SysconfTest.java:
```diff
@@ -0,0 +1,70 @@
+            tickrate = sysconf.main();
+        }
+        catch (IOException e) {
+            System.err.println(e);
```
Try/catch should not be used, AssertDoesNotThrow instead

Laukkala committed:
Removed "Throws" keywords from tests and replaced with an additional Assertion that tests do not throw any exception

Laukkala committed:
Addded pre-generated compiled C files so that SysconfTest can be run with no preparation.
These files can be generated into /target folder via command "mvn install"

Laukkala committed:
Spotless formatting changes

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,252 @@
+
+            Assertions.assertTrue(stat.ctxt() >= 0);
+            Assertions.assertTrue(stat.intr().size() >= 0);
+            Assertions.assertTrue(stat.processes() >= 0);
```
Can some of these values actually ever be zero or negative?

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/LinuxOSTest.java:
```diff
@@ -0,0 +1,252 @@
+    public void highLevelMethodsTest() {
+        Assertions.assertDoesNotThrow(() -> {
+            LinuxOS os = new LinuxOS();
+            Assertions.assertTrue(Integer.signum(os.cpuCount()) == 1);
```
Maybe assertEquals `==` comparisons just for clarity

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,355 @@
+        ArrayList<Integer> pids = new ArrayList<Integer>();
+        Assertions.assertDoesNotThrow(() -> {
+
+            java.lang.Process ps = Runtime.getRuntime().exec("ps");
```
I'd generally immediately reject any external calls in tests but I guess here it is unavoidable. `ps` is not guaranteed to be installed so it might fail on containers, can you add rextra check that it actually exists?

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,355 @@
+        ArrayList<Integer> pids = new ArrayList<Integer>();
+        Assertions.assertDoesNotThrow(() -> {
+
+            java.lang.Process ps = Runtime.getRuntime().exec("ps");
```
Is getting the pids with `pidof` actually better?
```
$ pidof emacs
110477
$ pidof firefox
109655 108572 99249 98876 81986 9076 9072 8998 6397 6254 6207 5697 5674 5663 5655 5354 5318 5294 5158
```

Laukkala replied:
Will "pidof" have the same problem as with "ps", it might not be installed on containers?

StrongestNumber9 replied:
Yes, not guaranteed to exist

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,355 @@
+
+    // Kthreadd process (pid = 1) should have 52 fields in its Stat file, and it should have (systemd) as comm value and 1 as pid value in the file. None of the fields should be null
+    @Test
+    public void readSystemdFileTest() {
```
This test will fail in containers as systemd isn't available there

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,355 @@
+            TestThread processThread = new TestThread(pb);
+            processThread.run();
+
+            ArrayList<Integer> pids = getPidByCommand("sleep");
```
There can be other sleeps running on the system at the same time so this is can be fairly fragile test

Laukkala replied:
Preferably this test would create a process, make sure that isAlive() is true, then kill the process and verifiy that isAlive is now false. If there are other sleeps running, this test will fail, and worse yet, it will kill the sleep process it finds (even if it's someone else's process).

This could be solved if we knew the process ID of the created process, but java.lang.Process.pid() is a Java 9 feature. There seems to be some ways to get it via reflection, but at least in my tests I've not been able to access the process id field within Java's ProcessImpl easily.

Is there some way we can create a process where we are able to identify it properly without knowing its pid?
Or should we just test isAlive() by calling it on the JVM process, verifying that it returns true, and leaving it at that?
At least we shouldn't be killing sleep processes that have been started somewhere else.

Laukkala replied:
I think I managed to solve this:
We don't need to kill "sleep", since it terminates itself after the given time, we can just wait for the duration of the sleep process before checking isAlive() again.
Added a check that makes sure there's only one sleep process active. If there's more than one, the test will simply fail. Otherwise we might get false positives.

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,355 @@
+        });
+    }
+
+    // The number of tasks should be equal to the number of threads in /proc/pid/stat. Test for both systemd and the current JVM
```
Tests two things -> should be two different tests as the content of pid 1 and jvm stats do not depend on each other

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,355 @@
+            Statm statm = systemd.statm();
+
+            // Timestamps should always have a value
+            Assertions.assertTrue(stat.timestamp() != null);
```
Assertions.assertNotNull instead of `!= null` check

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,355 @@
+            Process jvm = new Process(pid, new LinuxOS(new SysconfInterface.Fake()));
+            float cpuTime1 = jvm.cpuTime();
+            ArrayList<Double> workArray = new ArrayList<Double>();
+            for (int i = 0; i < 30000000; i++) { // Needs to do a lot of work to have different CPUtimes since the times are in seconds, not ticks
```
Seems like this will take at least 240MB of memory, shouldn't be an issue though

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/SysconfTest.java:
```diff
@@ -0,0 +1,66 @@
+            tickrate = 0;
+            Sysconf sysconf = new Sysconf();
+            tickrate = sysconf.main();
+            System.out.println(tickrate);
```
Don't print in tests unless necessary

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/SysconfTest.java:
```diff
@@ -0,0 +1,66 @@
+            Sysconf sysconf = new Sysconf();
+            tickrate = sysconf.main();
+            System.out.println(tickrate);
+            Assertions.assertFalse(tickrate == 0 && tickrate == -1);
```
Maybe spit these to their own assertions

StrongestNumber9 commented on src/test/java/com.teragrep.jos_01.procfs/TaskTest.java:
```diff
@@ -0,0 +1,156 @@
+            Statm statm = task.statm();
+
+            // Timestamps should always have a value
+            Assertions.assertTrue(stat.timestamp() != null);
```
assertNotNull

StrongestNumber9 reviewed (changes requested):
Didn't read the code but added some notes that caught my eye in the tests, some are admittedly fairly nitpicky

Laukkala committed:
Added a check that makes sure that "ps" is installed when getting a process ID
Changed readSystemdFileTest to read the proc files of the JVM instead, as systemd might not exist within containers.
Split some tests into two to make tests more granular
Made some tests more readable

Laukkala committed:
Removed the "kill" command from isAliveTest, as there could be other "sleep" processes active at the time when the test is run. Instead the test will simply wait until enough time has elapsed so that the sleep process is not running anymore.
Added a check to isAliveTest() that makes sure that no other sleep processes are active while test is being run to avoid false positives.

Laukkala committed:
CpuTimeTest: Reduced the number of iterations when doing "work" by factor of 10 to reduce memory usage

Laukkala committed:
Fixed a bug where Process.cpuUsage() was rounding down processes' cpu usage when process time on cpu was short

Laukkala committed:
Updated readme with more detailed installation instructions

Laukkala committed:
Updated gitignore

Laukkala committed:
Removed old gitignore

Laukkala committed:
Added memoryPercentage() to Process, which will return the percentage of system memory being used by the process.

Laukkala committed:
Formatting changes for memoryPercentage()

Laukkala committed:
Made SysconfInterface.Fake() have a configurable tickrate instead of hardcoded 100

Laukkala committed:
Updated Readme

Laukkala committed:
Updated Readme

Laukkala committed:
Added a constructor to LinuxOS accepting only a proc directory path for ease of use

Laukkala committed:
Applied Exception chaining pattern to objects throwing Exceptions

Laukkala committed:
Updated Readme

kortemik commented:
exceptions catching hides the original cause, i.e. try { read() } catch (Exception e) { throw new Exception("some error"); } does not indicate that read() actually failed. this is sub-optimal.

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
-This is a template repository for Teragrep organization.
+This project will extract statistics from the operating system and running processes into Java objects.
+You can use this to easily get information such as memory usage about running processes and the operating system itself.
+The information provided is based on the Linux filesystem's /proc files.
```
change /proc to `/proc`

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
 == Limitations
 
-// If your project has limitations, please list them. Otherwise remove this section.
+Only supports Linux-based operating systems
```
Use proper sentences or make an unordered list

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
 == Features
 
-// List your project's features
+Check if a process is alive using process id.
```
- who checks? now it assumes the user checks if a process is alive
- id -> ID

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
 
-// List your project's features
+Check if a process is alive using process id.
+Get statistics on the OS or any running process using your system's /proc files as the source of truth.
```
/proc -> `/proc`

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
-// List your project's features
+Check if a process is alive using process id.
+Get statistics on the OS or any running process using your system's /proc files as the source of truth.
+High level functions calculate process performance statistics automatically, such as residentSetSize()
```
- A period is missing
- residentSetSize() -> `residentSetSize()`

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
+Check if a process is alive using process id.
+Get statistics on the OS or any running process using your system's /proc files as the source of truth.
+High level functions calculate process performance statistics automatically, such as residentSetSize()
+Create objects representing the current state of a /proc file and get access to the file's values.
```
- who creates?
- /proc -> `/proc`

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
+You can use this to easily get information such as memory usage about running processes and the operating system itself.
+The information provided is based on the Linux filesystem's /proc files.
 
 == Features
```
This section seems like it should be an unordered list but it's not. Add one or rewrite the section in the way it's a consistent paragraph and not list of random sentences.

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
 
-// If your project has limitations, please list them. Otherwise remove this section.
+Only supports Linux-based operating systems
+Relies on availability of /proc files, so JOS-01 needs to have access to these files.
```
/proc -> `/proc`

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
+Only supports Linux-based operating systems
+Relies on availability of /proc files, so JOS-01 needs to have access to these files.
 
 == How to [compile/use/implement]
```
remove brackets and use one of following titles:

- How to Compile
- How to Use
- How to Implement

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
 == How to [compile/use/implement]
 
-// add instructions how people can start to use your project
+Requirements:
```
Use sub title and unordered list

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
+Java version 1.8
+Maven
+
+Installation:
```
use sub title

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
+
+Installation:
+Include the following dependency into your project's pom.xml file:
+    <dependency>
```
use code block

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
+      <artifactId>jos_01</artifactId>
+    </dependency>
+
+// Run "mvn clean install" successfully before use.
```
This is currently commented out, either use remove or use a code block

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
+// - autoreconf
+// - gcc
+
+Usage:
```
Use sub title

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
+// - gcc
+
+Usage:
+Operating system statistics:
```
Use sub title

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
+Optionally provide an object implementing SysconfInterface to manually configure the Operating system CPU tick rate.
+Use methods of LinuxOS object to get information on operating system
+
+Process statistics:
```
Use sub title

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
-== Contributing
 
-// Change the repository name in the issues link to match with your project's name
+Examples of use:
```
Use sub title

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
+Examples of use:
+
+Creating a new LinuxOS object with automatic configuration:
+LinuxOS os = new LinuxOS();
```
Use code block

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,86 @@
+LinuxOS os = new LinuxOS();
+
+Creating a new Process object with automatic configuration:
+Process process = new Process(250);
```
Use code block

Laukkala committed:
Updated Readme

Laukkala commented:
> exceptions catching hides the original cause, i.e. try { read() } catch (Exception e) { throw new Exception("some error"); } does not indicate that read() actually failed. this is sub-optimal.

The underlying Exceptions should still be present, since the exception is included within the second parameter of the re-thrown exception:

```
catch (Exception e) { 
	throw new Exception("some error", e);
}
```
The underlying exceptions can be printed out with Throwable.printStackTrace(), or accessed with Throwable.getCause()
I think most loggers should print out the causes automatically and end up with something like the exception below:
```
java.lang.Exception: Failed to calculate system CPU count!
	at com.teragrep.jos_01.procfs.LinuxOS.cpuCount(LinuxOS.java:139)
Caused by: java.lang.Exception: Failed to create a Cpuinfo object!
	at com.teragrep.jos_01.procfs.LinuxOS.cpuinfo(LinuxOS.java:168)
	... 71 more
Caused by: java.io.FileNotFoundException: wontfind/cpuinfo (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
```

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
 == Features
 
-// List your project's features
+- Can be used to check if a process is alive using process ID.
```
"...is alive BY using process ID."

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
 
-// List your project's features
+- Can be used to check if a process is alive using process ID.
+- Provides statistics on the OS or any running process using your system's `/proc` files as the source of truth.
```
"Provides the statistics of OS or any running process by using your..."

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+== How to use
 
-// add instructions how people can start to use your project
+## Requirements
```
Change ## to === to get h3 level title

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+- Maven
 
-== Contributing
+## Installation
```
Change ## to === to get h3 level title

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+      <artifactId>jos_01</artifactId>
+    </dependency>
+
+## Usage
```
Change ## to === to get h3 level title and add line break after

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+    </dependency>
+
+## Usage
+### Operating system statistics
```
Change ### to ==== to get h4 level title

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+Optionally provide an object implementing SysconfInterface to manually configure the Operating system CPU tick rate.
+Use methods of LinuxOS object to get information on operating system
+
+### Process statistics
```
Change ### to ==== to get h4 level title

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+Optionally provide a specific LinuxOS object during construction so that OS configuration is applied to Process object's methods
+
+
+## Examples of use
```
Change ## to === to get h3 level title

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+## Examples of use
 
-// Change the repository name in the issues link to match with your project's name
+### Creating a new LinuxOS object with automatic configuration
```
Change ### to ==== to get h4 level title and add line break

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+### Creating a new LinuxOS object with automatic configuration
+    LinuxOS os = new LinuxOS();
+
+### Creating a new Process object with automatic configuration
```
Change ### to ==== to get h4 level title and add line break

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+### Creating a new Process object with automatic configuration
+    Process process = new Process(250);
+
+### Creating a new LinuxOS object with manual CPU tickrate configuration
```
Change ### to ==== to get h4 level title and add line break

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+### Creating a new LinuxOS object with manual CPU tickrate configuration
+    LinuxOS os = new LinuxOS(SysconfInterface.Fake(100));
+
+### CPU count of current system:
```
Change ### to ==== to get h4 level title and add line break

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+    LinuxOS os = new LinuxOS();
+    System.out.println(os.cpuCount());
+
+### Manually configure CPU tickrate and get CpuTime of process with ID 1
```
Change ### to ==== to get h4 level title and add line break

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
+    Process process = new Process(1,os);
+    System.out.println(process.cpuTime());
+
+### Access /proc/stat files of process with ID 250 and print out "comm" value
```
Change ### to ==== to get h4 level title and add line break

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,79 @@
-== Contributing
+## Installation
+Include the following dependency into your project's pom.xml file:
+    <dependency>
```
Add line break before to get the code block. Markdown and asciidoc needs an empty line between paragraphs and element blocks, otherwise the text is printed on the same line.

Laukkala committed:
Updated Readme

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,87 @@
+=== Usage
+
+==== Operating system statistics
+Start by creating a LinuxOS object.
```
Add a line break between the title and first paragraph

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,87 @@
+==== Operating system statistics
+Start by creating a LinuxOS object.
+Optionally provide an object implementing SysconfInterface to manually configure the Operating system CPU tick rate.
+Use methods of LinuxOS object to get information on operating system
```
A period is missing from the sentence.

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,87 @@
+Use methods of LinuxOS object to get information on operating system
+
+==== Process statistics
+Create a Process object by supplying a valid process ID to Process' constructor.
```
Add a line breaker between a title and the first paragraph

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,87 @@
+
+==== Operating system statistics
+Start by creating a LinuxOS object.
+Optionally provide an object implementing SysconfInterface to manually configure the Operating system CPU tick rate.
```
"Optionally provide an object BY implementing..."

If SysconfInterface is a parameter or anything else code related, turn it into `SysconfInterface`

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,87 @@
+
+==== Process statistics
+Create a Process object by supplying a valid process ID to Process' constructor.
+Optionally provide a specific LinuxOS object during construction so that OS configuration is applied to Process object's methods
```
a comma is missing out between construction and so: "...during construction, so that..."

the sentence is missing out a period

Laukkala committed:
Updated Readme

ronja-ui commented on README.adoc:
```diff
@@ -1,35 +1,90 @@
+
+==== Process statistics
+Create a Process object by supplying a valid process ID to Process' constructor.
+Optionally provide a specific LinuxOS object during construction, so that OS configuration is applied to Process object's methods
```
A period is missing from the sentence

Laukkala committed:
Updated Readme

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Sysconf.java:
```diff
@@ -0,0 +1,85 @@
+    public interface LibSysconf extends Library {
+
+        LibSysconf INSTANCE = (LibSysconf) Native
+                .load("/lib/opt/Fail-Safe/jos_01/lib/sysconf/sysconf.so", LibSysconf.class);
```
rename to /opt/teragrep

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/Sysconf.java:
```diff
@@ -0,0 +1,85 @@
+        }
+        catch (UnsatisfiedLinkError unsatisfiedLinkError) {
+            throw new Exception(
+                    "Failed to initialize Native C library! Make sure that classes/lib/opt/Fail-Safe/jos_01/lib/sysconf/sysconf.so exists and is readable! Try running \"mvn clean install\" to generate required files",
```
rename to /opt/teragrep

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/SysconfInterface.java:
```diff
@@ -0,0 +1,70 @@
+// Exposes a Fake class that can be used for testing without having to compile the C code. Fake class simply returns a hardcoded value
+public interface SysconfInterface {
+
+    final class Fake implements SysconfInterface {
```
Fakes must be part of /test directory so they are not part of the final bundle.

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/status/PlainText.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.ArrayList;
+import java.util.Collections;
+
+public class PlainText implements Text {
```
is this PlainText or TimeAddedText or similar?

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/status/PlainText.java:
```diff
@@ -0,0 +1,79 @@
+public class PlainText implements Text {
+
+    private final ArrayList<String> fields;
+    private final LocalDateTime timestamp;
```
Please use Instant instead of LocalDateTime. See Procedure: Code Programming.

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/status/RowFile.java:
```diff
@@ -0,0 +1,95 @@
+public class RowFile extends File implements Text {
+
+    private final BufferedReader reader;
+    private final LocalDateTime timestamp;
```
Please use Instant instead of LocalDateTime. See Procedure: Code Programming.

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/status/Text.java:
```diff
@@ -0,0 +1,56 @@
+
+    ArrayList<String> read() throws Exception;
+
+    LocalDateTime timestamp();
```
Please use Instant instead of LocalDateTime. See Procedure: Code Programming.

kortemik commented on src/test/resources/lib/opt/Fail-Safe/jos_01/lib/sysconf/sysconf.la:
```diff
@@ -0,0 +1,41 @@
+dlpreopen=''
+
+# Directory that this library needs to be installed in:
+libdir='/opt/Fail-Safe/jos_01/lib/sysconf'
```
rename to /opt/teragrep

kortemik reviewed (changes requested):
add src/test/resources/lib/opt/Fail-Safe/jos_01/lib/sysconf/sysconf.so to git ignore as well, a compiled binary must not be part of the repository.

Laukkala committed:
Refactored timestamps to use Instants instead of LocalDateTime objects.

Laukkala committed:
Spotless formatting changes

Laukkala committed:
Renamed Sysconf library directory from 'Fail-Safe' to 'teragrep'

Laukkala committed:
Renamed PlainText to TimeaddedText as the primary responsibility of the object is to represent text that has been instantiated at a certain time

Laukkala committed:
Moved SysconfInterface.Fake into its' own class file within /test directory so that it will not be included in the final build.

Laukkala committed:
Spotless formatting changes

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/status/process/Stat.java:
```diff
@@ -0,0 +1,396 @@
+    }
+
+    public void printTimestamp() {
+        LOGGER.info(Date.from(timestamp).toString());
```
Date.from formats as 
```
dow mon dd hh:mm:ss zzz yyyy
```

which is in conflict with Procedure: Code Programming which states that ISO 8601 is a good format. Please change these

kortemik replied:
I.e. do not use java.util.Date, use instead Instant.toString() which provides ISO 8601.

Laukkala replied:
Removed usage of Date.from

Laukkala committed:
printTimestamp() now uses ISO8601 format for timestamps

kortemik commented on src/main/java/com/teragrep/jos_01/procfs/status/os/Cpuinfo.java:
```diff
@@ -0,0 +1,365 @@
+            processorFields.add(value);
+        }
+        processors.add(new Processor(processorFields));
+        this.timestamp = origin.timestamp();
```
please do not copy origin.timestamp() into a field in the decorator class, instead reference to it via origin.timestamp() whenever needed.

Laukkala committed:
Removed code from constructors by calling origin in timestamp() methods

kortemik commented on src/test/resources/lib/opt/Fail-Safe/jos_01/lib/sysconf/sysconf.la:
```diff
@@ -0,0 +1,41 @@
+# sysconf.la - a libtool library file
```
built artifact must not be part of the repository. please configure tests to use the built artifact from the target/ directory. please see https://maven.apache.org/plugins-archives/maven-surefire-plugin-2.12.4/examples/configuring-classpath.html for more information

q22u commented on src/main/java/com/teragrep/jos_01/procfs/ProcFile.java:
```diff
@@ -0,0 +1,78 @@
+            System.out.println("Reason: " + notFoundException.getMessage());
+        }
+        catch (IOException ioException) {
+            System.out.println("Failed to read file \"" + this.getName() + "\" due to an I/O exception!");
```
Remove method getName

Laukkala replied:
This seems to be an outdated file

q22u commented on src/main/java/com/teragrep/jos_01/procfs/ProcFile.java:
```diff
@@ -0,0 +1,78 @@
+        }
+        catch (IOException ioException) {
+            System.out.println("Failed to read file \"" + this.getName() + "\" due to an I/O exception!");
+            System.out.println("Reason: " + ioException.getMessage());
```
Consider other method for message

Laukkala replied:
This seems to be an outdated file

q22u commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,142 @@
+    public ArrayList<Task> tasks() {
+        ArrayList<Task> tasks = new ArrayList<Task>();
+        for (File taskDirectory : new File(procDirectory, "task").listFiles()) {
+            tasks.add(new Task(Long.parseLong(taskDirectory.getName()), this));
```
Reconsider using the getmethod

Laukkala replied:
File.getName() is from Java's IO library, so the name not really object oriented, unless there is some easy alternative?

q22u commented on src/main/java/com/teragrep/jos_01/procfs/Task.java:
```diff
@@ -0,0 +1,135 @@
+        return fileNames;
+    }
+
+    private ArrayList<String> procFileNames(ArrayList<String> nameList, File file) {
```
This seems weird way to get file names, why do we have alot of file names and paths? maybe this list could be done in other way or clarified

Laukkala replied:
I think we can just remove this method entirely, it was used in a previous prototype but that has since been refactored and this method no longer has any uses.

q22u commented on src/main/java/com/teragrep/jos_01/procfs/Task.java:
```diff
@@ -0,0 +1,135 @@
+        else {
+            try {
+                for (File child : file.listFiles()) {
+                    procFileNames(nameList, child);
```
What is file child

Laukkala replied:
I think we can just remove this method entirely, it was used in a previous prototype but that has since been refactored and this method no longer has any uses.

q22u commented on src/main/java/com/teragrep/jos_01/procfs/Process.java:
```diff
@@ -0,0 +1,142 @@
+                }
+            }
+            catch (NullPointerException npe) {
+                System.out.println("I/O Exception while attempting to access children of " + file.getPath());
```
Are we sure about this catch?

Laukkala replied:
This one seems to be an outdated file, catches for NullPointerExceptions have been taken out

q22u commented on src/test/test.java/com.teragrep.jos_01.procfs/TaskTest.java:
```diff
@@ -0,0 +1,77 @@
+        ArrayList<Task> tasks = process.tasks();
+        Map<String, String> statistics = tasks.get(0).stat().statistics();
+        Assert.assertEquals(52, statistics.size());
+        Assert.assertEquals("0", statistics.get("ppid"));
```
maybe add more tests for statistics as we have several options

Laukkala replied:
This seems to be an old version of the test file, the current one has tests for both stat() and statm(), which are the only statistics methods available for Task objects for now

q22u commented on src/test/test.java/com.teragrep.jos_01.procfs/ProcessTest.java:
```diff
@@ -0,0 +1,116 @@
+        rows = process.proc("comm").rows();
+        Assert.assertEquals(1, rows.size());
+        Assert.assertEquals("systemd", rows.get(0));
+    }
```
Maybe we could consider adding test to see procDirectory and procFile give us expected instances?

Laukkala replied:
I found and fixed a bug in Process' constructor thanks to this! Process's procDirectory field was being hardcoded to "/proc" instead of using the same as LinuxOS object. I also added a test case for this bug.

q22u commented on src/main/java/com/teragrep/jos_01/procfs/status/process/Statm.java:
```diff
@@ -0,0 +1,129 @@
+// Provides information about memory usage of the process.
+// Fields are on a single row, separated by spacebars. Contains 7 fields.
+// All fields are integers.
+public class Statm implements Text {
```
Instead of commenting, may we have proper object name StatMemory / stats memory

Laukkala replied:
This object should be an instance of Text in the defined format of the Statm proc file, which is why I named it Statm. The format is defined in man pages: https://man7.org/linux/man-pages/man5/proc_pid_statm.5.html

q22u commented on src/main/java/com/teragrep/jos_01/procfs/status/process/Statm.java:
```diff
@@ -0,0 +1,129 @@
+    private final long shared;
+    private final long text;
+    private final long lib;
+    private final long data;
```
May we know why this class is named data and what they should tell about memory to us?

Laukkala replied:
It's the memory usage of the processes' data and stack, in pages.
I chose these names because  they are used in the man pages for this proc file:
https://man7.org/linux/man-pages/man5/proc_pid_statm.5.html

Laukkala committed:
Removed C output files from test/resources and configured maven to get them from target instead

Laukkala committed:
Fixed a bug where instantiating a Process object without specifying a /proc directory would use hardcoded "/proc" path instead of using the same directory as the OS object by default
Added a unit test for this case

Laukkala committed:
Removed unused method procFileNames()

Laukkala committed:
Removed unnecessary call to System.out.println() from test
